### PR TITLE
Implement new lint for detecting buggy pointer-to-int casts

### LINKED ISSUE #81686: Emit warning when using `as` to directly cast function pointers to any integral type other than usize
Imagine you have a `u32`. You want to check if it's larger than the maximum value of a `u16`. Of course, you know that the API will give you a `u16`, so you will need to cast it to a `u32`.

You write a unit test. 100 million should be above the range of a `u16`, you think.
```rust
let x = 100_000_000;
let y = u16::max as u32;
println!("{}", x > y);
```
Output:
```
false
```
You begin to doubt your computer science education. In a fugue, you abandon all worldly possessions. You move to the woods. You raise chickens now; sometimes, though, you wonder if they're truly the ones raising you. It is a simpler life.

---

`u16::max` is the UFCS'd form of the default'd `max` method on the prelude'd `Ord` trait. To a first approximation, there is no reason to ever directly cast a function pointer to anything other than a `usize`. The compiler should emit a warning when attempting to cast a function pointer to `u8`, `u16`, `u32`, `u64`, `u128`, `i8`, `i16`, `i32`, `i64`, `i128`, and `isize`. This would not apply to transitive casts; `foo as usize as i8` would not warn.

(If I had realized this existed, I would have renamed `u16::MAX` [back when I had the chance](https://github.com/rust-lang/rfcs/pull/2700).)

### PR DESCRIPTION
**Pull request by osa1**
Fixes #81686

### CODE REVIEW

**Comment by varkor** [comment]
@osa1: are you waiting for this to be reviewed, or is it not ready yet? (I assumed you were still working on it because it was marked as a draft, but now I realise this might not be the case.)

**Comment by osa1** [comment]
Hi @varkor -- sorry for late reply. This is ready for reviews.

Note that the lint is not enabled by default, I'm guessing we want to enable it by default before shipping, right?

I realize I didn't add a test -- I'll do that now.

File: compiler/rustc_typeck/src/check/expr.rs
Comment (osa1) [inline compiler/rustc_typeck/src/check/expr.rs:1051]: I should split this into multiple lines somehow.. this is currently too long for a single line.

File: compiler/rustc_typeck/src/check/expr.rs
Comment (varkor) [inline compiler/rustc_typeck/src/check/expr.rs:1044]:
```suggestion
                        let msg = format!("casting pointer to `{}`", t_cast);
```

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (varkor) [inline compiler/rustc_lint_defs/src/builtin.rs:3001]:
```suggestion
    "detects pointers casts to integer types other than `usize`",
```

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (varkor) [inline compiler/rustc_lint_defs/src/builtin.rs:2999]: I think `INVALID_PTR_TO_INT_CAST` or similar would be a slight improvement, as `usize` is also an integer.

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (varkor) [inline compiler/rustc_lint_defs/src/builtin.rs:2970]:
```suggestion
    /// The `ptr_to_int_cast` lint triggers if a pointer is cast to any integer type other than
    /// `usize`, since doing so is often a bug.
```

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (varkor) [inline compiler/rustc_lint_defs/src/builtin.rs:2979]:
```suggestion
    ///     let y = u16::max as u32; // the user meant the constant `u16::MAX`, rather than the
    ///                              // function `u16::max`, but this cast is technically permitted,
    ///                              // so will not produce an error
```

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (varkor) [inline compiler/rustc_lint_defs/src/builtin.rs:2980]:
```suggestion
    ///     println!("{}", x > y); // prints `false` (unexpectedly)
```

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (varkor) [inline compiler/rustc_lint_defs/src/builtin.rs:2990]:
```suggestion
    /// The example above shows how a user might accidentally cast a function pointer (rather
    /// than an integer constant) to a non-`usize` integer type. Though this is currently permitted
    /// by Rust, there is very little reason to cast a pointer to any integer other than a `usize`.
    /// Therefore, any instances of such casts are likely to be bugs.
```

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (varkor) [inline compiler/rustc_lint_defs/src/builtin.rs:2993]:
```suggestion
    /// This lint warns against those cases.
```

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (varkor) [inline compiler/rustc_lint_defs/src/builtin.rs:2995]:
```suggestion
    /// In the future, we may want to make this a hard error.
```

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (varkor) [inline compiler/rustc_lint_defs/src/builtin.rs:2998]:
```suggestion
    /// To cast a pointer to a non-`usize` integer without triggering the lint, you can first cast
    /// to a `usize` and then to the integer type, e.g. `ptr as usize as u32`.
```

File: compiler/rustc_typeck/src/check/expr.rs
Comment (varkor) [inline compiler/rustc_typeck/src/check/expr.rs:1037]:
```suggestion
            let is_integral_cast = matches!(t_cast.kind(), ty::Int(_) | ty::Uint(_));
```
It's not possible to cast pointers to floats anyway.

File: compiler/rustc_typeck/src/check/expr.rs
Comment (varkor) [inline compiler/rustc_typeck/src/check/expr.rs:1046]:
```suggestion
```
No point repeating the message.

File: compiler/rustc_typeck/src/check/expr.rs
Comment (varkor) [inline compiler/rustc_typeck/src/check/expr.rs:1047]:
```suggestion
                            .help(
```
I think this works for lints too, and it's more appropriate than a label if so.

File: compiler/rustc_typeck/src/check/expr.rs
Comment (varkor) [inline compiler/rustc_typeck/src/check/expr.rs:1049]:
```suggestion
                                "pointers should only be cast to `usize`",
```

File: compiler/rustc_typeck/src/check/expr.rs
Comment (varkor) [inline compiler/rustc_typeck/src/check/expr.rs:1057]:
```suggestion
                            .span_suggestion(
                                expr.span,
                                &format!("to cast to `{}`, cast to `usize` first", t_cast),
                                &format!("{} as usize as {}", e.span, t_cast),
                                Applicability::MaybeIncorrect,
                            )
```
This may not quite be valid, but I think something similar would be more helpful.

File: src/test/ui/lint/lint-ptr-to-int-cast.rs
Comment (varkor) [inline src/test/ui/lint/lint-ptr-to-int-cast.rs:8]: Could you also add some tests for signed integer types and `isize`?

**Review by varkor — CHANGES_REQUESTED** [review]
Thanks, this looks pretty good so far! I just have a few comments about wording, to try to make it a little clearer for users.

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (varkor) [inline compiler/rustc_lint_defs/src/builtin.rs:3000]: > Note that the lint is not enabled by default, I'm guessing we want to enable it by default before shipping, right?

I think it would be reasonable to make this `Warn` by default. We'll need to get the lang team to sign off on the new lint.

**Comment by osa1** [comment]
Hm, it seems like the check is not quite right. This program doesn't generate a warning:

```rust
fn main() {
    fn test() {}
    let x: u16 = ({
        test
    }) as u16;
}
```

**Comment by osa1** [comment]
> Hm, it seems like the check is not quite right. This program doesn't generate a warning:

Figured it out, update coming.

**Comment by osa1** [comment]
@varkor should be ready now.

**Comment by osa1** [comment]
Ping @varkor

File: compiler/rustc_typeck/src/check/cast.rs
Comment (varkor) [inline compiler/rustc_typeck/src/check/cast.rs:585]:
```suggestion
                                Applicability::MaybeIncorrect,
```
This presence of this lint means that we expect the user not to have intended to write the cast, so likely the user will want to rewrite their code, rather than ignore the lint by casting via `usize`.

**Review by varkor — COMMENTED** [review]
I think this is mostly looking good now. Could you address the one comment, and squash your commits (or tidy them up a bit, e.g. there shouldn't be fixup or merge commits)?

**Comment by osa1** [comment]
@varkor all done.

**Comment by varkor** [comment]
@rust-lang/lang: this PR adds a new lint that triggers when a pointer is cast to a non-`usize` integer. The intention is to reduce easy-to-miss bugs like `u16::max as u32` (rather than `u16::MAX as u32`): additional motivation is described in https://github.com/rust-lang/rust/issues/81686.

At the moment, this lint is allow-by-default, but I suggest it makes sense to make it warn-by-default. If this seems reasonable, could someone start an FCP for the new lint?

**Review by varkor — APPROVED** [review]

**Comment by scottmcm** [comment]
A warning for these cases makes good sense to me 👍 

A few thoughts:

- Would it make sense to split the lints for `fn`s and ordinary `*const`s?  This feels substantially worse for casts from function items (like the `i32::max_value as u32` example) than for casts from data pointers, though I agree even for normal pointers a cast to anything but `[ui]size` is suspicious at best.

- I think this should not use the `invalid` term in the lint name.  The other use of that is inline with Ralf's terminology in the [`invalid-value`](https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#invalid-value) lint, but here the cast, while likely undesirable, is still _valid_.

**Comment by joshtriplett** [comment]
I'd be happy to see this become warn-by-default for most types. I'd prefer not to see a warning by default for casts to `u64`, though. There are valid use cases for using a fixed-size field to hold a pointer, rather than using a `usize` which varies in size across targets.

`u64` aside, I'd otherwise be happy to see this become a warning.

**Comment by osa1** [comment]
(rebased)

**Comment by nikomatsakis** [comment]
Nominating for our lang team meeting based on [@varkor's comment](https://github.com/rust-lang/rust/pull/81789#issuecomment-788712012)

**Comment by Mark-Simulacrum** [comment]
> I'd be happy to see this become warn-by-default for most types. I'd prefer not to see a warning by default for casts to `u64`, though. There are valid use cases for using a fixed-size field to hold a pointer, rather than using a `usize` which varies in size across targets.
> 
> `u64` aside, I'd otherwise be happy to see this become a warning.

I think I'd personally still prefer to see this as `as usize as u64` or so, such that the widening (or potential truncation, if on a platform with more than 64 bit pointers, though I don't think we have any today) is explicit. At the very least, we shouldn't allow this if we added >64-bit usize targets at any point.

**Comment by cramertj** [comment]
We discussed this in the most recent @rust-lang/lang meeting and agreed with @joshtriplett's [comment above](https://github.com/rust-lang/rust/pull/81789#issuecomment-790127762) that we'd prefer not to lint on pointer-to-`u64` casts, as practically these are likely to be correct and portable, and we have a general policy not to allow warn-by-default lints to contain false positives.

**Comment by osa1** [comment]
I rebased the branch and made a commit to allow casts to u64. To make this easier to review the change is done in a new commit. I'll squash before merging.

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (joshtriplett) [inline compiler/rustc_lint_defs/src/builtin.rs:2954]: I think it'd be fine to say "you can first cast to a `usize` and then" here; that will give people guidance if they don't already have a specific need.

The rest of the change seems fine.

File: compiler/rustc_typeck/src/check/cast.rs
Comment (joshtriplett) [inline compiler/rustc_typeck/src/check/cast.rs:587]: Likewise, here, I think it's fine to just say "cast to `usize` first". We don't need to give two different recommendations.

**Comment by joshtriplett** [comment]
Two items above, where I think it'd be fine to give specific guidance steering people towards `usize`. The rest of the change seems fine.

With that added, I'll propose this for FCP to confirm lang-team consensus.

**Comment by osa1** [comment]
- Addressed comments
- Rebased

I will squash commits in the final version.

Just a reminder: the lint is not on by default. Do we want to change that, or wait for lang team's feedback on this?

**Comment by joshtriplett** [comment]
@osa1 I'd personally be in favor of making this warn-by-default. (I'd also like to see the output of a crater run for that, to get an idea of the impact on the ecosystem, and to see if there's some common pattern we're missing that does this. Would also be good to get out ahead of this and fix any major crates supplying macros that do this.)

**Comment by osa1** [comment]
(rebased)

**Comment by osa1** [comment]
I made the lint deny-by-default, for the crater run.

**Comment by osa1** [comment]
Can I skip the compiler tests before the crater run somehow? I don't want to update all these tests just to run crater with the lint is denied by default.

**Comment by bjorn3** [comment]
The failed job is one of the PR CI jobs. Try builds still succeed if tests are failing as far as I know.

**Comment by bjorn3** [comment]
@craterbot check

**Comment by osa1** [comment]
3 days and the job is still in the queue? :astonished:

**Comment by varkor** [comment]
@osa1: crater takes a while to complete (it has to compile all the crates on cargo.io), and the queue is quite busy at the moment.

**Comment by cramertj** [comment]
@rust-lang/lang discussed this again recently. We felt fairly hesitant about adding this lint.

When we discussed previously, there was consensus that we did *not* want to lint on pointer -> `u64` casts. However, at that time we didn't properly understand the main problem this lint was aiming to solve. After reviewing [the original issue](https://github.com/rust-lang/rust/issues/81686), it seems like the lint introduced in this PR doesn't solve the general problem described in that issue.

There are a number of related issues occurring:
1. I meant to call a function and cast the result but I cast a function pointer instead. Example: `u16::max as u32` when I wanted `u16::max() as u32`.
2. I meant reference a constant and cast it but I cast a similarly named function pointer instead. Example: `u16::max as u32` when I wanted `u16::MAX as u32`.
3. I meant to dereference a pointer and cast the result but I cast the pointer instead. Example: `ptr as u32` when I wanted `*ptr as u32`.
4. I accidentally wrote non-portable code by casting a pointer to `u32`. Example: `ptr as u32` when I wanted to store pointers in an integer, possibly `ptr as u64`.

As far as I can tell, it seems like the primary issue described in https://github.com/rust-lang/rust/issues/81686 is (1) and maybe (2). This PR seems to primarily address (4), with the side effect that some (but not all) instances of 1-3 are also detected.

It feels like the lint in this PR as-implemented is likely to introduce a number of false positives (I really did want to store that pointer in a u32), yet a number of the cases in the original bug are not detected (e.g. `u16::max as u128`).

Personally, I feel like this issue is another reason why we should offer convenient and ergonomic replacements for `as` (e.g. `my_u8.upcast():u64`, `my_u64.truncate():u8`), at which point we could guide users away from `as` entirely, possibly even removing it on an edition boundary.

**Comment by osa1** [comment]
I think #81686 describes (2), not (1). Quoting the original issue: 

> The compiler should emit a warning when attempting to cast a function item to u8, u16, u32, u64, u128, i8, i16, i32, i64, i128, and isize

However, what the issue describes may not matter too much, we may still want to provide a lint for casting function items/pointers to integer types when it will obviously overflow. We already do this for some other cases, for example:

```rust
fn main() {
    let x: u32 = u32::MAX + 1;
}
```

generates

```
error: this arithmetic operation will overflow
 --> src/main.rs:2:18
  |
2 |     let x: u32 = u32::MAX + 1;
  |                  ^^^^^^^^^^^^ attempt to compute `u32::MAX + 1_u32`, which would overflow
  |
  = note: `#[deny(arithmetic_overflow)]` on by default
```

It seems inconsistent to not do the same check when casting a pointer to a smaller integer type.

> It feels like the lint in this PR as-implemented is likely to introduce a number of false positives (I really did want to store that pointer in a u32)

Could you say more about your use case? Is this on a 32-bit platform?

> yet a number of the cases in the original bug are not detected (e.g. u16::max as u128).

We could detect this case as well. It was originally detected, but removed per @joshtriplett's request: https://github.com/rust-lang/rust/pull/81789#issuecomment-790127762

**Comment by osa1** [comment]
As I say above, I think this lint makes sense regardless what you think #81686 is about, but if the language team disagrees then please feel free to close this PR, or let me know and I will close it.

**Comment by scottmcm** [comment]
We discussed this again in the @rust-lang/lang meeting this week.  We didn't come to a clear consensus, though, so what I'm posting here is a proposal for a direction to take this, not anything final or signed-off.

I propose that we move forward with this lint, but *remove* the "Fixes #81686" from the OP.  That would mean that this lint becomes just about general pointer truncation, and another PR would target that issue more specifically.

That means that *this* lint will *not* solve the `isize::max as usize` problem.  I would propose that that be addressed by an additional PR that would focus on casts from *function voldemort ZSTs* specifically, where we can lint more aggressively than we can on pointers.  (We certainly don't want to lint on `as usize` on a normal pointer -- although maybe we could move that way in the future after some hypothetical libs additions that could be used instead to be specific in intent.)

I think that also implies a few changes here.  For example, the crater run shows a bunch of errors from casting to `isize` in order to subtract, which seems to me like they probably don't meet the threshold for a rustc lint.

**Comment by nikomatsakis** [comment]
I'm curious about this statement by @cramertj:

> It feels like the lint in this PR as-implemented is likely to introduce a number of false positives (I really did want to store that pointer in a u32)

This seems like the important question! I can imagine this comes up because of portability (e.g., when targeting 32-bit systems), but I'm not sure if there are other times when people really do want to store pointers into a u32 (which of course remains an option, but more awkward).

**Comment by Mark-Simulacrum** [comment]
We [discussed](https://github.com/rust-lang/lang-team/blob/master/minutes/2021-05-11.md#implement-new-lint-for-detecting-buggy-pointer-to-int-casts-rust81789) this in the last lang team meeting.

We noted that:

* We'd like to have a concrete example of a real-world bug that the lint *in this PR* (not described on the issue) is fixing
  * We've eliminated a lot of cases where the lint would fire already (u64/i64, isize/usize...)
* In terms of portability, if writing code targeting only 32-bit systems, it may be beneficial to have an allow-by-default lint catching additional cases.
  * We're pretty sure we *don't* want a lint that fires only on some targets, because that's confusing and hard to manage for users.
* We would like to have these lint(s) recommend a "better path", and suspect that better path is explicit to/from_bits methods, at least to start - a [thread has been kicked off on Zulip to discuss those](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Adding.20methods.20as.20more.20specific.20versions.20of.20.60as.60).
  * That would let us be more aggressive about deprecating parts of `as` casts, in favor of methods which are more explicit about exactly what the user wanted.

We didn't discuss next steps too concretely, but it feels like it might be good for someone to propose a PR adding to and from_bits to `*const T`, `*mut T`, and perhaps `fn(...)` and the zero-sized function types. For the first two some of the discussion in the lang meeting indicated we may want to restrict to just 'thin' pointers to start (via `<T as Pointee>::Metadata = ()`. For the latter that starts to touch on Niko's proposal in https://github.com/rust-lang/lang-team/issues/23.

r? @joshtriplett for now as you've been driving this the most I think

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/lang meeting, as well as last week's @rust-lang/libs meeting.

In @rust-lang/lang, our consensus was that we'd love to see `as` used much less, and we'd be happy to lint against it in cases where we have some specific function to recommend that better expresses the intent and *isn't* `as X as Y`. And if we get to the point where most functions of `as` are served by functions that better express intent, we could consider changing the behavior of `as` in a new edition (e.g. making it only a shorthand for a subset of type changes, like widening).

In @rust-lang/libs, our consensus was that we'd be happy to take proposals for functions that serve the function of `as` (e.g. `truncate` for integer narrowing, `to_bits` to turn a pointer into an integer, `to_signed` or `to_unsigned` for signedness changes only). We'd love to see a comprehensive effort in this regard, but we'd be happy to take individual proposals for such functions, as long as we feel they'd fit into a coherent overall design.

@rust-lang/libs also expressed the sentiment that `as` is a very convenient shorthand, especially since it doesn't require writing turbofish (`as u8` seems lexically simpler than `truncate::<u8>()`), but overall still seemed amenable to the idea of providing functions that better express intent.

We're going to close this PR, as we don't think we're going to accept this precisely as written. We'd be thrilled to see functions proposed to libs, and then corresponding lints proposed to lang to use those functions rather than `as`.

**Comment by joshtriplett** [comment]
To clarify: we'd be happy to see some version of this lint, and we think there's value in catching issues like this; we'd be happy to see this reopened and updated when there are methods that it can suggest.

**Pull request closed without merging**
